### DISCUSSION TITLE
Undesired changes to destination in compose options

### DISCUSSION THREAD

**Discussion by jayknott** [ImageMagick/ImageMagick#3524] [Help] [answered]
Hi community,

I have been trying to fix this for a week now and have run out of Google articles to read. When doing alpha compositing with other compose options besides src-over, there are changes in the destination image even if the source image is xc:none. Let me walk through a simple recreation of the issue and hopefully some brilliant user will be able to help or explain this to me.

Using art_01.png as my destination and xc:none as my source.

art_01.png (just random stripes with semi-transparency):
![art_01](https://user-images.githubusercontent.com/24280037/114225464-306a9e00-9927-11eb-9986-f50bd91c7507.png)

```
magick convert -size 500x500 art_01.png xc:none -compose over -composite over.png
```

over.png looks fine; no changes to the destination whatsoever:
![over](https://user-images.githubusercontent.com/24280037/114225982-d8806700-9927-11eb-8c5b-f34c1f8e514f.png)

When I compare semitransparent pixels, they are the same:
```
magick convert art_01.png -format "%[pixel:p{30,70}]" info:
srgba(37,239,10,0.494118)
magick convert over.png -format "%[pixel:p{30,70}]" info:
srgba(37,239,10,0.494118)
```

Now, using a compose option of dst, which should completely ignore the source and keep the destination the same, I am seeing changes being made to the image.

```
magick convert -size 500x500 art_01.png xc:none -compose dst -composite dst.png
```

dst.png:
![dst](https://user-images.githubusercontent.com/24280037/114226480-783df500-9928-11eb-88e2-45581be71a2b.png)

As you can see the "blur" around the lines has been reduced. I have no idea what is causing this change as according to the documentation it should not happen. This is happening with most compose options other than src-over.

When comparing pixels, you can see that there is a change:

```
magick convert art_01.png -format "%[pixel:p{30,70}]" info:
srgba(37,239,10,0.494118)
magick convert dst.png -format "%[pixel:p{30,70}]" info:
srgba(18,118,5,0.494118)
```

I am running CLI for Windows and here is my build info:
```
magick -version
Version: ImageMagick 7.0.11-6 Q16 x64 2021-04-03 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Visual C++: 192829912
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zip zlib
```

I am just trying to figure out how to prevent these changes or the cause of them so I can correct it. Any help would be appreciated.

**Comment by fmw42** [comment]
First a couple of comments. With IM 7, use magick, not magick convert. The latter will use IM 6 syntax. Second, -size, in general, should be set just before xc:, though won't make a difference  here. 

If you look at https://legacy.imagemagick.org/Usage/compose/#dst, it seems to indicate that the alpha channel of the source image will be disabled. That will leave a black background image to which the dst image as overlay will blend. So fully transparent regions will stay fully transparent and opaque regions will stay opaque. But partially transparent regions will mix with black. So I tested

```
magick  art_01.png -size 500x500 xc:none -compose dst -composite dst1.png
magick  art_01.png -size 500x500 xc:black -compose dst -composite dst2.png
magick compare -metric rmse -channel rgba dst1.png dst2.png null:
0 (0)

```
which validates my interpretation.

**Reply by jayknott** [reply]
Thanks for the reply. I did try not using convert as well, but it had the same result. What you are saying makes sense. Do you have a suggestion on how I can accomplish what I am looking to do?

**Reply by jayknott** [reply]
In the end I am trying to do a compose a multiply or screen later with the destination. But even if I do a multiply with just white which should have no change, it still has this effect.

**Reply by fmw42** [reply]
What is the command you are trying to do?  With what image? Do the images have alpha channels?

**Reply by jayknott** [reply]
I am basically trying to add shadow to an image. Here is an example:

art.png is the destination image:
![blue_04_image_only](https://user-images.githubusercontent.com/24280037/114237427-f0f87d80-9937-11eb-9c6b-09db780821ca.png)

shirt.png is the source image:
![02_normal_transparent](https://user-images.githubusercontent.com/24280037/114237486-040b4d80-9938-11eb-8ba5-b20b09be1580.png)

```
magick -size 1200x1200 art.png shirt.png -compose multiply -composite mult.png
```

![mult](https://user-images.githubusercontent.com/24280037/114238476-847e7e00-9939-11eb-8983-80bfa54ed9ed.png)

As you can see, when using multiply, the semi-transparent areas are mixed with white.

I know that I can just swap these images and multiply this on top of the shirt image, but that won't always work because sometimes the base shirt will be another color other than white. I need to add the shadows using multiple and then I compose this on top of another colored shirt. Both of these have alpha channels and need to stay that way.  Even if I try it with xc:white or -channel rgb it still changes art.png semi-transparent color values. I obviously don't understand how these calculations work and need help. I am trying to mimic how a multiply layer would work in Photoshop.

**Reply by jayknott** [reply]
I am also willing to purchase some personal consulting with someone who knows ImageMagick in and out.

**Comment by snibgo** [comment]
IM tries to comply with SVG compositing operations, https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_dst

I think IM v7 does comply. For "-compose dst", the output alpha is set to the "dst" ("destination") alpha, and the output colour is set to the "dst" colour multiplied by the "dst" alpha.

Beware of statements like:
> ... a compose option of dst, which should completely ignore the source and keep the destination the same, ...

That statement assumes dst has no transparency.

**Comment by snibgo** [comment]
For this task, "overlay" can be used. Depending on the order of the arguments, we get one of these:
![x1](https://user-images.githubusercontent.com/33812368/114242731-4c039200-9983-11eb-81eb-ae62183f0f86.png)
![x2](https://user-images.githubusercontent.com/33812368/114242740-4f971900-9983-11eb-837c-c0176942a543.png)

**Reply by jayknott** [reply]
I need to have the shadow transferred to the art, the shirt shouldn't be visible in the final image.

**Reply by fmw42** [reply]
Please show an example of exactly what you want.

**Reply by jayknott** [reply]
It's just like this one from the previous thread, but this result had white mixed in the semi-transparent pixels.

![image](https://user-images.githubusercontent.com/24280037/114244556-97964b80-9943-11eb-8399-03b7185743db.png)

**Comment by snibgo** [comment]
I still don't understand what you want.

You have two images:

1. A tee-shirt with transparent background.

2. Coloured marks, fading to a transparent background.

What output do you want? Perhaps you want transparency exactly as from image (2), so mostly transparent, and the colours (eg a and b channels from Lab colorspace) from (2) but the lightness from (1).

Windows BAT syntax:
```
magick ^
  ( 114237486-040b4d80-9938-11eb-8ba5-b20b09be1580.png ^
    -colorspace Lab ^
    -channel 0 -separate +channel ^
  ) ^
  ( 114237427-f0f87d80-9937-11eb-9c6b-09db780821ca.png ^
    ( +clone -alpha extract -write mpr:ALPH +delete ) ^
    -colorspace Lab ^
    -channel 1,2  -separate +channel ^
  ) ^
  -combine ^
  -set colorspace Lab -colorspace sRGB ^
  mpr:ALPH ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  x5.png
```
![x5](https://user-images.githubusercontent.com/33812368/114259419-ade4eb80-99c5-11eb-82b5-c223a1e73c1a.png)

**Comment by jayknott** [comment]
@fmw42  and @snibgo,

Through your questions and explanations I have realized what I am asking for is not possible. Well, not in the way I thought it was. And everything is working as it should. I am working out a new method to accomplish what I want with my new knowledge. I genuinely thank you for jumping on and helping so quickly.

I don't know if credit matters for this forum, but if someone wants to post something stating it's not possible for what I am trying to do I will mark it as the answer.

**Comment by snibgo** [comment, marked as answer]
I don't know if what you want is possible, because I don't know what you want.

You can mark your own reply as "the answer" if you want.

A variation on my previous is to multiply the RGB channels of the coloured marks by the lightness of the tee-shirt. As the shirt lightness is less than 1 (on a scale of 0 to 1), this will darken all the colours.
```
magick ^
  ( 114237427-f0f87d80-9937-11eb-9c6b-09db780821ca.png ^
    ( +clone -alpha extract -write mpr:ALPH +delete ) ^
  ) ^
  ( 114237486-040b4d80-9938-11eb-8ba5-b20b09be1580.png ^
    -colorspace Gray ^
  ) ^
  -alpha off ^
  -compose Multiply -composite ^
  mpr:ALPH ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  x6.png
```

![x6](https://user-images.githubusercontent.com/33812368/114272352-c24cd680-9a0d-11eb-9c24-1a07607fcb5e.png)

**Reply by jayknott** [reply]
This is actually exactly what I want! I wanted to just darken the image and keep the alpha layer the same.

**Comment by snibgo** [comment]
Good stuff.

Note that I have used the lightness of the shirt, by converting it to grayscale. In this example, the shirt is already gray. If the shirt was some other colour, the command would still work, transferring the shadows (the lightness) from the shirt.
